### DISCUSSION TITLE
Changes to `HexaPDF::Layout` classes

### DISCUSSION THREAD

**Discussion by gettalong** [gettalong/hexapdf#185] [General]
Hi!

There will be some fairly major changes in `HexaPDF::Composer` and the whole `HexaPDF::Layout` module in the upcoming release. Among others, there is a new `ColumnBox` class for laying out content in columns.

Additionally, **I want to make it easier to use the classes from the `HexaPDF::Layout` module without using `HexaPDF::Composer`.**

Therefore I would like to ask if there are pain points when using those layout classes and if so, what are the pain points?

**Comment by rubyFeedback** [comment]
I have not yet used either variant; my use cases so far are still quite limited.

I guess looking at the names only, I would assume ColumnBox to be a bit like the 2-div layout in newspapers or 3-div layouts, e. g. side by side? My own long term goal is to be able to render all my autogenerated HTML pages (I use ruby to autogenerate the code actually) as .pdf files eventually seamlessly. There are work arounds for this, e. g. headless rendering and stuff like that, but I'd kind of love to be able to do the "write once, generate everything", so being able to offer .pdf for all my webpage content one day would be great, without having to spend too much time styling any of that (I already do the styling on the webpage, so it would be nice if .pdf could take some of that. Does not have to be perfect, just the basic stuff, colours, borders, spacing, margin ... that's it.)

**Reply by gettalong** [reply]
Yes, the `ColumnBox` allows one to layout content in one or more columns, with automatic height balancing if needed. See e.g. https://twitter.com/_gettalong/status/1534468170295267328.

As for your goal: There is no HTML-to-PDF converter for HexaPDF and none is planned at the moment. This is very difficult to do properly and would be a project on its own. If you need something like this, I would recommend https://weasyprint.org/.

What there will be in the not too distant future is a kramdown-to-PDF converter based on HexaPDF. The only missing part for it is support for simple tables in HexaPDF. Currently the PDF converter is based on Prawn and works fine. I have used it in the past to generate various versions of content automatically for a lecture, see https://www.mat.univie.ac.at/~praxis/wisem15/vorlesung.html
